File: python/cross_db.py
#!/usr/bin/env python3

import psycopg2
import configparser
import sys
import xml.etree.ElementTree as ET
import json

def read_config(config_file):
    """
    Reads the configuration file and returns the database credentials
    """
    config = configparser.ConfigParser()
    config.read(config_file)
    return {
        "host": config.get("postgresql", "host"),
        "port": config.getint("postgresql", "port"),
        "user": config.get("postgresql", "user"),
        "password": config.get("postgresql", "password")
    }

def get_databases(config):
    """
    Connects to the PostgreSQL server and returns a list of all databases
    """
    try:
        with psycopg2.connect(
            host=config["host"],
            port=config["port"],
            user=config["user"],
            password=config["password"],
            dbname="postgres"
        ) as conn:
            with conn.cursor() as cur:
                cur.execute("SELECT datname FROM pg_database WHERE datistemplate = false;")
                return [db[0] for db in cur.fetchall()]
    except (Exception, psycopg2.Error) as error:
        print(f"Error while connecting to PostgreSQL: {error}")
        return None

def run_query(config, sql_file, db):
    """
    Connects to the PostgreSQL server and runs the query on the specified database
    """
    try:
        with psycopg2.connect(
            host=config["host"],
            port=config["port"],
            user=config["user"],
            password=config["password"],
            dbname=db
        ) as conn:
            with conn.cursor() as cur:
                cur.execute(open(sql_file, 'r').read())
                return (db, cur.fetchall())
    except (Exception, psycopg2.Error) as error:
        print(f"Error while connecting to PostgreSQL: {error}")

from xml.dom import minidom

def write_output_to_xml_file(output, file):
    """
    Writes the query output to the specified file in a pretty XML format.
    """
    root = ET.Element("databases")
    for (db, rows) in output:
        db_element = ET.SubElement(root, "database", name=db)
        for row in rows:
            ET.SubElement(db_element, "row").text = str(row)

    rough_string = ET.tostring(root, 'utf-8')
    reparsed = minidom.parseString(rough_string)
    with open(file, 'w') as f:
        f.write(reparsed.toprettyxml(indent="  "))

def write_output_to_json_file(output, file):
    """
    Writes the query output to the specified file in a pretty JSON format.
    """
    data = {}
    for (db, rows) in output:
        data[db] = [row for row in rows]

    with open(file, 'w') as f:
        json.dump(data, f, indent=2)

def write_output_to_file(output, file):
    """
    Writes the query output to the specified file in a more readable format
    """
    with open(file, 'w') as f:
        for (db, rows) in output:
            f.write(f'Results for {db}:\n')
            for row in rows:
                f.write(str(row) + '\n')
            f.write('\n')

if __name__ == "__main__":
    if len(sys.argv) < 4:
        print("Usage: python script.py config.ini query.sql output.out [xml|json|txt]")
        sys.exit()

    if len(sys.argv) == 4:
        config_file, sql_file, output_file = sys.argv[1:]
        output_format = "xml"

    if len(sys.argv) == 5:
        config_file, sql_file, output_file, output_format = sys.argv[1:]
    
    config = read_config(config_file)
    databases = get_databases(config)
    output = []
    for db in databases:
        output.append(run_query(config, sql_file, db))
    if output_format == "txt":
        write_output_to_file(output, output_file)
    elif output_format == "xml":
        write_output_to_xml_file(output, output_file)
    elif output_format == "json":
        write_output_to_json_file(output, output_file)

File: rust/cross_db.rs
extern crate postgres;
extern crate config;

use postgres::{Client, NoTls};
use std::fs::File;
use std::io::prelude::*;
use config::{Config, File as ConfigFile};

fn read_config(config_file: &str) -> config::Config {
    let mut settings = Config::new();
    settings.merge(ConfigFile::with_adjustments(config_file, |config| {
        config.merge(config::File::new("config", config::FileFormat::Ini))
    })).unwrap();
    return settings;
}

fn get_databases(config: &config::Config) -> Vec<String> {
    let conn_str = format!("postgres://{}:{}@{}:{}",
        config.get_str("postgresql.user").unwrap(),
        config.get_str("postgresql.password").unwrap(),
        config.get_str("postgresql.host").unwrap(),
        config.get_int("postgresql.port").unwrap());

    let mut client = Client::connect(&conn_str, NoTls).unwrap();
    let rows = client.query("SELECT datname FROM pg_database WHERE datistemplate = false;", &[]).unwrap();
    return rows.iter().map(|row| row.get(0)).map(|s| s.to_string()).collect();
}

fn run_query(config: &config::Config, sql_file: &str, db: &str) -> (String, Vec<Vec<postgres::types::ToSql>>) {
    let conn_str = format!("postgres://{}:{}@{}:{}/{}",
        config.get_str("postgresql.user").unwrap(),
        config.get_str("postgresql.password").unwrap(),
        config.get_str("postgresql.host").unwrap(),
        config.get_int("postgresql.port").unwrap(),
        db);

    let mut client = Client::connect(&conn_str, NoTls).unwrap();
    let mut sql_file = File::open(sql_file).unwrap();
    let mut sql = String::new();
    sql_file.read_to_string(&mut sql).unwrap();
    let rows = client.query(&sql, &[]).unwrap();
    return (db.to_string(), rows);
}

fn write_output_to_file(output: Vec<(String, Vec<Vec<postgres::types::ToSql>>)>, file: &str) {
    let mut file = File::create(file).unwrap();
    for (db, rows) in output {
        file.write_all(format!("Results for {}:\n", db).as_bytes()).unwrap();
        for row in rows {
            for val in row {
                file.write_all(format!("{} ", val).as_bytes()).unwrap();
            }
            file.write_all("\n".as_bytes()).unwrap();
        }
        file.write_all("\n".as_bytes()).unwrap();
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 4 {
        println!("Usage: script config.ini query.sql output.txt");
        std::process::exit(1);
    }
    let config_file = &args[1];
    let sql_file = &args[2];
    let output_file = &args[3];
    let config = read_config(config_file);
    let databases = get_databases(&config);
    let mut output = vec![];
    for db in &databases {
        output.push(run_query(&config, sql_file, db));
    }
    write_output_to_file(output, output_file);
}

